### PR TITLE
Add tutorial for arbitrary precision in ADNLPModels

### PR DESCRIPTION
**Pull request by tmigot**
**New tutorial**
Tutorial name: Multiprecision and genericity in NLPModels with automatic differentiation

### CODE REVIEW

**Review by abelsiqueira — CHANGES_REQUESTED** [review]
Thanks for the new tutorial, Tangi. I had some issues understanding what was the focus of the tutorial because I did not get the issue that this was trying to solve. I made some comments and suggestions to try to make it clear.
One issue, that is not this tutorial's problem, is that even if we create the multi-precision capable model, we still can't use it in a multi-precision algorithm, i.e., `lbfgs` with `nlp{Float64}` and `x::Float32`.

File: tutorials/generic-adnlpmodels/index.jmd [hunk 1]
```diff
@@ -0,0 +1,133 @@
+---
+title: "Multiprecision and genericity in NLPModels with automatic differentiation"
+tags: ["models", "automatic differentiation", "multi-precision", "tests"]
+author: "Tangi Migot"
+---
+
+
+# Multiprecision and genericity in NLPModels with automatic differentiation
```
[the remark this answers is not in the document]
Comment (tmigot) [reply tutorials/generic-adnlpmodels/index.jmd]: I changed it for `Creating an ADNLPModels backend that supports multiple types`

File: tutorials/generic-adnlpmodels/index.jmd [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+```julia
+using ADNLPModels, ForwardDiff, NLPModels, OptimizationProblems
+```
+
+One of the main strengths of Julia for scientific computing is its native usage of [arbitrary precision arithmetic](https://docs.julialang.org/en/v1/manual/integers-and-floating-point-numbers/#Arbitrary-Precision-Arithmetic).
+
+The same can be exploited for optimization models and solvers.
+In the organization [JuliaSmoothOptimizers](https://juliasmoothoptimizers.github.io), the package [ADNLPModels.jl](https://github.com/JuliaSmoothOptimizers/ADNLPModels.jl) provides automatic differentiation (AD)-based model implementations that conform to the NLPModels API.
+This package is modular in the sense that it implements a backend system allowing the user to use essentially any AD system available, see [ADNLPModels.jl/dev/backend/](https://juliasmoothoptimizers.github.io/ADNLPModels.jl/dev/backend/) for a tutorial.
```
[the remark this answers is not in the document]
Comment (tmigot) [reply tutorials/generic-adnlpmodels/index.jmd:18]: Sure, it is possible

**Comment by tmigot** [comment]
Thanks for the comments @abelsiqueira . 

I added a more detailed introduction to the problem.

About the solvers, it is mainly a JSO-internal issue right now. Several students/postdocs are working on multi-precision algorithms and asked for this feature.

**Review by abelsiqueira — APPROVED** [review]
Thanks for the changes.

**Comment by d-monnet** [comment]
Thank you for the tutorial, simple and efficient.

I only have a few comments:
- In the [tutorial](https://juliasmoothoptimizers.github.io/JSOTutorials.jl/generic-adnlpmodels/), the Rosenbrock function and initial point are not interpreted correctly.
- Maybe it would worth splitting the  Multiprecision ADNLPModels in two subsections for the sake of clarity: Objective Evaluation and Gradient Evaluation. This would help to emphasized that multi precision evaluation of objective and gradient relies on two different mechanisms (generic typing of x0 and backend).
- It would be nice to have the counters to record the number of eval for each precision that has been used. Not sure how much of a hassle it would be to implement that though.
- In your last example, 
  ```julia
  x16 = Float16.(get_x0(nlp))
  ```
  is typically something I would avoid to do. Casting x0 from Float64 into Float16 implies that rounding error occur, therefore x0 is different than x16 and the gradient is not evaluated as x0. I would keep the example as it is, but point out that issue to the user and mention that one should be extra careful when using multi precision.

**Comment by tmigot** [comment]
Thanks @d-monnet for the comments, I updated the tuto !

**Comment by tmigot** [comment]
@d-monnet about counters for multi-precision, I think it would be too much changes for now as I don't really see a simple way. Because right now thé `incrément!` function doesn't have access to thé type.

**Pull request merged**